<?xml version="1.0"?>
<launch>

    <!-- rviz config -->
    <!-- <arg name="rviz_config" default="-d $(find px4_sim)/config/config.rviz" /> -->
    <arg name="rviz_config" default="-d $(find px4_sim)/config/slam_2d.rviz" />

    <!-- rviz 显示 -->
    <!-- <node pkg="joint_state_publisher" type="joint_state_publisher" name="joint_state_publisher" /> -->
    <node name="rviz_2d" pkg="rviz" type="rviz" args="$(arg rviz_config)" />

    <!-- 里程计真值 -->
    <node pkg="px4_sim" type="odmo_gt_pub" name="odmo_gt_pub" />

    <!-- gmapping 2D SLAM -->
    <param name="use_sim_time" value="true"/>
    <node pkg="gmapping" type="slam_gmapping" name="slam_gmapping" output="screen">
      <remap from="scan" to="laser_scan"/> <!-- 雷达话题 -->
      <remap from="map" to="gmapping_map"/> <!-- 地图话题 -->
      <remap from="odmo" to="odmo_2d" />
      <param name="base_frame" value="base_link"/> <!-- 底盘坐标系 -->
      <param name="odom_frame" value="odmo_2d"/> <!-- 里程计坐标系 -->
      <param name="map_update_interval" value="5.0"/>
      <param name="maxUrange" value="16.0"/>
      <param name="sigma" value="0.05"/>
      <param name="kernelSize" value="1"/>
      <param name="lstep" value="0.05"/>
      <param name="astep" value="0.05"/>
      <param name="iterations" value="5"/>
      <param name="lsigma" value="0.075"/>
      <param name="ogain" value="3.0"/>
      <param name="lskip" value="0"/>
      <param name="srr" value="0.1"/>
      <param name="srt" value="0.2"/>
      <param name="str" value="0.1"/>
      <param name="stt" value="0.2"/>
      <param name="linearUpdate" value="1.0"/>
      <param name="angularUpdate" value="0.5"/>
      <param name="temporalUpdate" value="3.0"/>
      <param name="resampleThreshold" value="0.5"/>
      <param name="particles" value="30"/>
      <param name="xmin" value="-50.0"/>
      <param name="ymin" value="-50.0"/>
      <param name="xmax" value="50.0"/>
      <param name="ymax" value="50.0"/>
      <param name="delta" value="0.05"/>
      <param name="llsamplerange" value="0.01"/>
      <param name="llsamplestep" value="0.01"/>
      <param name="lasamplerange" value="0.005"/>
      <param name="lasamplestep" value="0.005"/>
    </node>

    <!-- path planning -->
    <node pkg="move_base" type="move_base" respawn="false" name="move_base" output="screen" clear_params="true">
        <rosparam file="$(find px4_sim)/config/move_base/costmap_common_params.yaml" command="load" ns="global_costmap" />
        <rosparam file="$(find px4_sim)/config/move_base/costmap_common_params.yaml" command="load" ns="local_costmap" />
        <rosparam file="$(find px4_sim)/config/move_base/global_costmap_params.yaml" command="load" />
        <rosparam file="$(find px4_sim)/config/move_base/local_costmap_params.yaml" command="load" />
        <rosparam file="$(find px4_sim)/config/move_base/base_local_planner_params.yaml" command="load" />
        <remap from="scan" to="laser_scan"/> <!-- 雷达话题 -->
        <remap from="map" to="gmapping_map"/> <!-- 地图话题 -->
        <remap from="odmo" to="odmo_2d" />
    </node>
    
</launch>
